021-server-side-rendering/021-server-side-rendering.md

```diff
@@ -0,0 +1,296 @@
+
+**Problem #2: Accessing OAuth2 sessions**
+
+Here's a visualisation of the current OAuth2 flow:
+
+![CSR OAuth2 Flow Sequence Diagram](csr-oauth-flow.png)
+
+Step by step:
+
+1. Client makes a 'Create OAuth2 Session' request to the Appwrite,containing the provider, and a success authentication redirect URL.
+2. Appwrite returns a URL for a page to authenticate with the OAuth2 provider.
+3. User is redirected to the authentication URL, and authenticates with the OAuth2 provider.
+4. OAuth2 provider redirects the browser back to Appwrite.
+5. Appwrite sets the session cookie on the Appwrite domain, and redirects the browser to the success URL.
+
+This is incompatible with SSR applications, because the session cookie is set on the Appwrite domain, and not the SSR domain.
+
+> There is an undocumented workaround for SSR. To use it, when creating an OAuth2 session, set success parameter is set to `{SSR_DOMAIN}/auth/oauth2/success`. 
+> Now, Appwrite will append the session secret as a query parameter when redirecting to this URL. You can find the source code for this [here](https://github.com/appwrite/appwrite/blob/3f3d518f3664bcab281ee00b45dd2f2d387ffc72/app/controllers/api/account.php#L870).
+> Although this workaround has good developer experience, it is not secure. The session secret is exposed in the URL, and can be intercepted.
+
+**Problem #3: Using session tokens with the SDK**
+
+After acquiring a session token, you want to make authenticated requests. For example, getting the user's account details. Currently, to set a session token on the server-side, we have to set the `X-Fallback-Cookies` header.
+
+```js
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const serialisedCookies = JSON.stringify({
+  a_session_PROJECT_ID: sessionToken,
+});
+
+client.headers["X-Fallback-Cookies"] = serialisedCookies;
+
+const account = new Account(client);
+const currentUser = await account.get();
+```
+
+This is not intuitive, and requires the developer to understanding details of Appwrite that don't need to be exposed.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+**Problem #1: Solution A - Return session token in response body**
+
+Return the token within the session object, resulting in the follow example code:
+
+```js
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const account = new Account(client);
+const session = await account.createAnonymousSession();
+const sessionToken = session.token;
+```
```

Comment:
This works, but what do we then do with the sessionToken? How would a more complete flow look like?

For reference, [this is what I do now](https://github.com/TGlide/pluck-svelte/blob/main/src/lib/appwrite/account.server.ts) to achieve SSR in auth.